pkg/routing/signal.go

```diff
@@ -94,49 +93,248 @@ func (r *signalClient) StartParticipantSignal(
+	GetSeq() uint64
+}
+
+type SignalMessageWriter[SendType RelaySignalMessage] interface {
+	WriteOne(seq uint64, msg proto.Message) SendType
+	WriteMany(seq uint64, msgs []proto.Message) SendType
+}
+
+type SignalMessageReader[RecvType RelaySignalMessage] interface {
+	Read(msg RecvType) ([]proto.Message, error)
+}
+
+func CopySignalStreamToMessageChannel[SendType, RecvType RelaySignalMessage](
+	stream psrpc.Stream[SendType, RecvType],
+	ch *MessageChannel,
+	reader SignalMessageReader[RecvType],
+	config config.SignalRelayConfig,
+) error {
+	r := &signalMessageReader[SendType, RecvType]{
+		reader: reader,
+		config: config,
+	}
+	for msg := range stream.Channel() {
+		var res []proto.Message
+		res, err := r.Read(msg)
+		if err != nil {
+			return err
+		}
+		for _, r := range res {
+			if err = ch.WriteMessage(r); err != nil {
+				return err
 			}
-			for _, res := range msg.Responses {
-				if err = resChan.WriteMessage(res); err != nil {
-					break
-				}
+		}
+	}
+	return stream.Err()
+}
+
+type signalMessageReader[SendType, RecvType RelaySignalMessage] struct {
+	seq    uint64
+	reader SignalMessageReader[RecvType]
+	config config.SignalRelayConfig
+}
+
+func (r *signalMessageReader[SendType, RecvType]) Read(msg RecvType) ([]proto.Message, error) {
+	res, err := r.reader.Read(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.config.MinVersion >= 1 {
+		if r.seq < msg.GetSeq() {
+			return nil, errors.New("participant signal message dropped")
+		}
+		if r.seq < msg.GetSeq() {
+			n := int(msg.GetSeq() - r.seq)
+			if n > len(res) {
+				n = len(res)
 			}
+			res = res[n:]
 		}
+		r.seq += uint64(len(res))
+	}
+	return res, nil
+}
 
-		logger.Debugw("participant signal stream closed",
-			"error", err,
-			"room", ss.RoomName,
-			"participant", ss.Identity,
-			"connectionID", connectionID,
-		)
+type SignalSinkParams[SendType, RecvType RelaySignalMessage] struct {
+	Stream         psrpc.Stream[SendType, RecvType]
+	Logger         logger.Logger
+	Config         config.SignalRelayConfig
+	Writer         SignalMessageWriter[SendType]
+	CloseOnFailure bool
+}
 
-		resChan.Close()
-	}()
+func NewSignalMessageSink[SendType, RecvType RelaySignalMessage](params SignalSinkParams[SendType, RecvType]) MessageSink {
+	return &signalMessageSink[SendType, RecvType]{
+		SignalSinkParams: params,
+	}
+}
+
+var ErrSignalFailed = errors.New("signal stream failed")
+
+type signalMessageSink[SendType, RecvType RelaySignalMessage] struct {
+	SignalSinkParams[SendType, RecvType]
 
-	return connectionID, &relaySignalRequestSink{stream}, resChan, nil
+	mu       sync.Mutex
+	seq      uint64
+	queue    []proto.Message
+	writing  bool
+	draining bool
 }
 
-type relaySignalRequestSink struct {
-	psrpc.ClientStream[*rpc.RelaySignalRequest, *rpc.RelaySignalResponse]
+func (s *signalMessageSink[SendType, RecvType]) Close() {
+	s.mu.Lock()
+	s.draining = true
+	if !s.writing {
+		s.Stream.Close(nil)
+	}
+	s.mu.Unlock()
+
+	<-s.Stream.Context().Done()
```

Comment:
should the message sink block here? is there any expectations that it should only return once fully flushed?

Reply:
it has to block otherwise when there are pending writes and rtcservice begins closing the stream close races with flushing. this causes `TestSignlePublisher` and `TestWebhooks` to fail